File: swara-learn/src/main/java/com/swara/ml/neural/layer/LstmLayer.java
package com.swara.ml.neural.layer;

import org.apache.commons.math3.linear.ArrayRealVector;
import org.apache.commons.math3.linear.RealVector;

/**
 * A neural layer with a long short term memory (LSTM). LSTM networks are a special class of
 * recurrent neural networks capable of learning long-term dependencies, because they are able to
 * selectively modify elements of their hidden state.
 */
public class LstmLayer extends Layer {

    private final ForwardLayer forget;
    private final ForwardLayer input;
    private final ForwardLayer output;
    private final ForwardLayer detect;

    private RealVector memory;

    /**
     *
     *
     */
    public LstmLayer(int inputs, int outputs) {
        super(inputs, outputs);
        this.input  = new ForwardLayer(inputs + outputs, outputs, ForwardLayer.LOGISTIC);
        this.forget = new ForwardLayer(inputs + outputs, outputs, ForwardLayer.LOGISTIC);
        this.output = new ForwardLayer(inputs + outputs, outputs, ForwardLayer.LOGISTIC);
        this.detect = new ForwardLayer(inputs + outputs, outputs, ForwardLayer.TANH);

        // Initialize the memory and set the last output to empty.
        this.memory = new ArrayRealVector(outputs);
        this.history().push(this.memory);
    }

    @Override
    public RealVector forward(RealVector input) {
        // Input is a concatenation of previous output and new input.
        final RealVector in = this.history().peek().append(input);

        // Update the memory as a linear combination of recalled memory and detected updates.
        final RealVector recall = this.forget.forward(in).ebeMultiply(this.memory);
        final RealVector update = this.input.forward(in).ebeMultiply(this.detect.forward(in));
        this.memory = recall.add(update);

        // The cell selectively outputs normalized elements of its memory.
        return this.history().push(this.output.forward(in).ebeMultiply(this.memory.map(Math::tanh)));
    }

    @Override
    public RealVector backward(RealVector error, double lrate) {
        // Gradient Output Gate = error * tanh(ct) = error * norm;
        // Error cT += error * ot * (1 - tanh^2(ct))
        // Gradient Forget = err memory * last_memory
        // Gradient Input  = err memory * detect output
        // Gradient Update = err memory * err input
        // Error cT-1 = gradient * forget

//        Memory error is a combination of output error and previous memory error. Gate errors come from output or memory error and are a result of the chain rule.

        throw new UnsupportedOperationException();
    }

}

File: swara-learn/src/main/scala/com/swara/learn/neural/layers/LstmLayer.scala
package com.swara.learn.neural.layers

import breeze.numerics._
import com.swara.learn.neural.Result

/**
 * Special thanks to: 
 * - http://colah.github.io/posts/2015-08-Understanding-LSTMs/
 * - http://arunmallya.github.io/writeups/nn/lstm/index.html#/
 
 * @param forget Forget gate (sigmoid); "forgets" elements of the state.
 * @param input Input gate (sigmoid); determines the parts of the state that receive updates.
 * @param output Output gate (sigmoid); determines the output of the layer.
 * @param detect Detect gate (tanh); "detects" new candidate values for the state.
 */
class LstmLayer(
  forget: ForwardLayer,
  input:  ForwardLayer,
  output: ForwardLayer,
  detect: ForwardLayer,
  var state:  Vector,
  var memory: Vector
) {

  def forward(x: Vector): Result[Vector, Array[Vector]] = {
    // Input is a concatenation of previous output and new input.
    val concat = Vector.vertcat(this.memory, x)

    // Update the state as a linear combination of recalled state and detected updates.
    val recall = this.forget.forward(concat) :* this.state
    val update = this.detect.forward(concat) :* this.input.forward(concat)
    this.state = recall :+ update

    // Selectively output normalized elements of its state.
    this.memory = this.output.forward(concat) :* tanh(this.state)
    this.memory
  }

  def backward(examples: Seq[(Vector, Vector)]): Seq[Vector] = {
    // Output Layer Error (o_t) = error_t :* tanh(state_t)
    // State Error (c_t) += error_t :* output_t :* (1 - tanh^2(state_t))
    // Input Layer Error (i_t) = c_t :* detect_t
    // Forget Layer Error (f_t) = c_t :* state_(t-1)
    // Detect Layer Error (d_t) = c_t :* input_t
    // Previous State Error (c_(t-1)) = c_t :* forget_t
  }

//  def forward(x: Vector): Vector = {
//    // Input is a concatenation of previous output and new input.
//    val concat = Vector.vertcat(this.memory, x)
//
//    // Update the state as a linear combination of recalled state and detected updates.
//    val recall = this.forget.forward(concat) :* this.state
//    val update = this.detect.forward(concat) :* this.input.forward(concat)
//    this.state = recall :+ update
//
//    // Selectively output normalized elements of its state.
//    this.memory = this.output.forward(concat) :* tanh(this.state)
//    this.memory
//  }
//
//  def backward(examples: Seq[(Vector, Vector)]): Seq[Vector] = {
//    // Output Layer Error (o_t) = error_t :* tanh(state_t)
//    // State Error (c_t) += error_t :* output_t :* (1 - tanh^2(state_t))
//    // Input Layer Error (i_t) = c_t :* detect_t
//    // Forget Layer Error (f_t) = c_t :* state_(t-1)
//    // Detect Layer Error (d_t) = c_t :* input_t
//    // Previous State Error (c_(t-1)) = c_t :* forget_t
//  }

}

object LstmLayer {

  def apply(in: Int, out: Int): LstmLayer =
    new LstmLayer(
      ForwardLayer.sigmoid(in + out, out),
      ForwardLayer.sigmoid(in + out, out),
      ForwardLayer.sigmoid(in + out, out),
      ForwardLayer.tanh(in + out, out),
      Vector.zeros(out),
      Vector.zeros(out)
    )
}
